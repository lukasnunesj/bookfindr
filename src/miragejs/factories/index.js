/*
 * Mirage JS guide on Factories: https://miragejs.com/docs/data-layer/factories
 */

import book from './book';

/*
 * factories are contained in a single object, that's why we
 * destructure what's coming from users and the same should
 * be done for all future factories
 */
export default {
    ...book,
};
