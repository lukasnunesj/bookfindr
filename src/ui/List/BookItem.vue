<template>
    <a
        :href="link"
        target="_blank"
        class="list-group-item list-group-item-action d-flex justify-content-between align-items-center"
    >
        <div class="flex-column">
            <h5>
                {{ title }}
            </h5>
            <p>
                <small>{{ authorsString }}</small>
            </p>
            <p>
                <small>{{ publishedDateYear }}</small>
            </p>
        </div>
        <div class="image-parent">
            <img
                :src="imgUrl ? imgUrl.smallThumbnail : noimage"
                class="book-thumb"
            />
        </div>
    </a>
</template>

<script>
export default {
    name: 'BookItem',
    props: {
        title: String,
        link: String,
        authors: Array,
        imgUrl: Object,
        publishedDate: String,
    },
    data: function () {
        return {
            noimage: '../../assets/img/noimg.jpeg',
        };
    },
    computed:{
        authorsString(){
            if (!this.authors) {
                return 'by unknown';
            }
            this.authors.forEach((element, index) => {
                this.authors[index] = element
                    .toLowerCase()
                    .replace(/\b\w/g, (l) => l.toUpperCase());
            });
            return 'by ' + this.authors.join(', ');
        },
        publishedDateYear(){
            if (!this.publishedDate) {
                return 'unknown publishing year';
            }
            const [year] = this.publishedDate.split('-');
            return 'published at ' + year;
        }
    },
};
</script>

<style scoped>
.book-thumb {
    width: 6vw;
    height: auto;
}
</style>
