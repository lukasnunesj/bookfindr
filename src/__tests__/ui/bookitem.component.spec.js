import { mount } from '@vue/test-utils';
import BookItem from '@/ui/List/BookItem';
import { makeServer } from '../../miragejs/server';

const mountBookItem = (override) => {
    const book = server.create('book', {
        volumeInfo:{
            title: 'Livro 1',
            authors: ['Author 1', 'Author 2'],
            publishedDate: '2020-02-02',
            imageLinks: {
                smallThumbnail: "http://books.google.com/books/content?id=H0taAAAAYAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api",
                thumbnail: "http://books.google.com/books/content?id=H0taAAAAYAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api"
            },
            infoLink: "http://books.google.com.br/books?id=H0taAAAAYAAJ&dq=a&hl=&source=gbs_api",
        ...override
        }
    });
    
    return {
        wrapper: mount(BookItem, {
            propsData: {
                title: book.volumeInfo.title,
                authors: book.volumeInfo.authors,
                publishedDate: book.volumeInfo.publishedDate,
                imgUrl: book.volumeInfo.imageLinks,
                link: book.volumeInfo.infoLink
            }
        }),
        book
    }
}

describe('BookItem - unit', () => {
    let server;
    beforeEach(() => {
        server = makeServer({ environment: 'test' });
    });

    afterEach(() => {
        server.shutdown();
    });

    it('should mount the component', () => {
        const {wrapper} = mountBookItem();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('should should mount the component with all props provided', () => {
        const { wrapper } = mountBookItem();

        expect(wrapper.vm).toBeDefined();
        expect(wrapper.text()).toContain('Livro 1');
        expect(wrapper.text()).toContain('by Author 1, Author 2');
        expect(wrapper.text()).toContain('published at 2020');
    });

    it('should should mount the component and show filler text when author and publishing date are not provided', () => {
        const { wrapper } = mountBookItem({
            authors: null,
            publishedDate: null
        });

        expect(wrapper.vm).toBeDefined();
        expect(wrapper.text()).toContain('Livro 1');
        expect(wrapper.text()).toContain('by unknown');
        expect(wrapper.text()).toContain('unknown publishing year');
    });
});