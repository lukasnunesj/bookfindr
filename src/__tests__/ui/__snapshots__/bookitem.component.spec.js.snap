// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookItem - unit should mount the component 1`] = `
<a
  class="list-group-item list-group-item-action d-flex justify-content-between align-items-center"
  href="http://books.google.com.br/books?id=H0taAAAAYAAJ&dq=a&hl=&source=gbs_api"
  target="_blank"
>
  <div
    class="flex-column"
  >
    <h5>
      
            Livro 1
        
    </h5>
     
    <p>
      <small>
        by Author 1, Author 2
      </small>
    </p>
     
    <p>
      <small>
        published at 2020
      </small>
    </p>
  </div>
   
  <div
    class="image-parent"
  >
    <img
      class="book-thumb"
      src="http://books.google.com/books/content?id=H0taAAAAYAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api"
    />
  </div>
</a>
`;
